drizzle-valibot/src/index.ts

```diff
@@ -38,58 +45,69 @@ import {
 	union,
+	type UnionIssue,
 	uuid,
 } from 'valibot';
 
-const literalSchema = union([string(), number(), boolean(), null_()]);
+// TODO: I think types test would be great for this file to make sure types are inferred correctly
 
-type Json = typeof jsonSchema;
-
-export const jsonSchema = union([literalSchema, array(any()), record(any())]);
+type Json = string | number | boolean | null | { [key: string]: Json } | Json[];
+type LiteralIssue = StringIssue | NumberIssue | BooleanIssue | NullIssue;
+type JsonIssue = LiteralIssue | ArrayIssue | RecordIssue;
+type JsonSchema = BaseSchema<Json, Json, UnionIssue<JsonIssue> | JsonIssue>;
+export const jsonSchema = lazy(() =>
+	union([string(), number(), boolean(), null_(), array(jsonSchema), record(string(), jsonSchema)])
+) as JsonSchema;
 
 type MapInsertColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
-> = TColumn['_']['notNull'] extends false ? OptionalSchema<NullableSchema<TType>>
-	: TColumn['_']['hasDefault'] extends true ? OptionalSchema<TType>
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
+> = TColumn['_']['notNull'] extends false ? OptionalSchema<NullableSchema<TType, never>, never>
+	: TColumn['_']['hasDefault'] extends true ? OptionalSchema<TType, never>
 	: TType;
 
 type MapSelectColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
-> = TColumn['_']['notNull'] extends false ? NullableSchema<TType> : TType;
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
+> = TColumn['_']['notNull'] extends false ? NullableSchema<TType, never> : TType;
 
 type MapColumnToValibot<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
 	TMode extends 'insert' | 'select',
 > = TMode extends 'insert' ? MapInsertColumnToValibot<TColumn, TType>
 	: MapSelectColumnToValibot<TColumn, TType>;
 
 type MaybeOptional<
 	TColumn extends Column,
-	TType extends BaseSchema<any, any>,
+	TType extends BaseSchema<unknown, unknown, BaseIssue<unknown>>,
 	TMode extends 'insert' | 'select',
 	TNoOptional extends boolean,
 > = TNoOptional extends true ? TType
 	: MapColumnToValibot<TColumn, TType, TMode>;
 
 type GetValibotType<TColumn extends Column> = TColumn['_']['dataType'] extends infer TDataType
 	? TDataType extends 'custom' ? AnySchema
-	: TDataType extends 'json' ? Json
+	: TDataType extends 'json' ? JsonSchema
 	: TColumn extends { enumValues: [string, ...string[]] }
-		? Equal<TColumn['enumValues'], [string, ...string[]]> extends true ? StringSchema
-		: PicklistSchema<TColumn['enumValues']>
+		// TODO: Can we somehow check `.length` to know if it's a string schema with or without a max length check?
+		? Equal<TColumn['enumValues'], [string, ...string[]]> extends true ? StringSchema<undefined> | SchemaWithPipe<[StringSchema<undefined>, MaxLengthAction<string, number, undefined>]>
+		: PicklistSchema<TColumn['enumValues'], undefined>
```

Comment:
Is adding 'readonly' to these lines necessary?

Line 102
```diff
-  : PicklistSchema<TColumn['enumValues'], undefined>
+  : PicklistSchema<Readonly<TColumn['enumValues']>, undefined>
```
Line 135
```diff
-  [K in keyof TColumns & string]: MaybeOptional<
+  readonly [K in keyof TColumns & string]: MaybeOptional<
```
Line 151
```diff
-  [K in keyof TTable['_']['columns']]: MaybeOptional<
+  readonly [K in keyof TTable['_']['columns']]: MaybeOptional<
```

Before:
![Screenshot 2024-06-10 at 15 59 33](https://github.com/drizzle-team/drizzle-orm/assets/31349069/8b66580d-77ce-4aaa-9496-759d1486d860)

After:
![Screenshot 2024-06-10 at 15 59 14](https://github.com/drizzle-team/drizzle-orm/assets/31349069/eaaf2243-4013-4c54-abba-c79913290036)

Reply:
I think `readonly` is not necessary and is only added by Valibot because we use TypeScript's `as const` feature for specific generic types.

Reply:
Thanks for clarifying